File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> answer;
    
    void solve(int pos, int target, vector<int>& nums, vector<int> path){
        if(target == 0){
            answer.push_back(path);
            return;
        }
        if(target < 0){
            return;
        }
        for (int i = pos; i < nums.size(); i++) {
            path.push_back(nums[i]);
            solve(i, target - nums[i], nums, path);
            path.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> path;
        solve(0, target, candidates, path);
        return answer;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def dfs(self, pos, candidates, target, subset, res):
        if target == 0:
            res.append(subset[:])
            return
        if target < 0:
            return
        for i in range(pos, len(candidates)):
            subset.append(candidates[i])
            self.dfs(i, candidates, target - candidates[i], subset, res)
            subset.pop()
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []
        self.dfs(0, candidates, target, subset, res)
        return res

''' 
At each index we have 2 options. Either keep the current element or skip the current element.
Make a recursion tree of this selection. We can select an element multiple times so one call 
with same index another call from next index.
'''

# class Solution:
#     def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
#         res = []
        
#         def dfs(i, target, subset):
#             if target == 0:
#                 res.append(subset)
#                 return
#             if target < 0 or i >= len(candidates):
#                 return
#             dfs(i, target - candidates[i], subset + [candidates[i]])
#             dfs(i + 1, target, subset)
        
#         dfs(0, target, [])
#         return res

'''
Time: O(2^T * K), T = Target value, K = Average length of the combinations
Space: O(K * x), hypothatical space complexity
'''
            
                    